src/Hpack/Config.hs

```diff
@@ -375,7 +393,15 @@ readPackageConfig file = do
     Left err -> return (Left err)
     Right config -> do
       dir <- takeDirectory <$> canonicalizePath file
-      Right <$> mkPackage dir config
+      result@(_, pkg) <- mkPackage dir config
+      case traverse validateExecutable $ packageExecutables pkg ++ packageTests pkg ++ packageBenchmarks pkg of
+        Left err' -> return (Left err')
+        Right _ -> return (Right result)
+      where
+        validateExecutable :: Section Executable Executable -> Either String ()
+        validateExecutable ex = case executableMain . sectionData $ ex of
+          Nothing -> Left "Missing main"
+          Just _ -> Right ()
```

Comment:
From what I tested, `Cabal` does not require `main` top-level, e.g. the following is a valid executable section:

```
executable foo
  if os(windows)
    buildable: False
  else
    hs-source-dirs: src, driver
    main-is: Main.hs
```

So just omitting `validateExecutable` is fine I guess.